ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,179 @@
+	// The agent will keep track of the highest sequence number received in both PayloadMessage and TaskManifestMessage.
+	// If the sequence number on TaskManifestMessage is lower to the latest known one, the TaskManifestMessage must be discarded.
+	// ACS will guarantee that the sequence number is indeed increasing and valid for this use case.
+	if manifestSeqNum < agentLatestSeqNum {
+		logger.Warn(fmt.Sprintf("Discarding task manifest message. "+
+			"Sequence number from manifest is less than Agent latest sequence number."), logger.Fields{
+			"ManifestSeqNum":    manifestSeqNum,
+			"AgentLatestSeqNum": agentLatestSeqNum,
+		})
+		return nil
+	}
+
+	// Save the task manifest's messageID and sequence number so that it can be compared
+	// by the taskStopVerificationAckResponder.
+	err := tmr.messageIDAccessor.SetMessageID(aws.StringValue(message.MessageId))
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest messageID")
+	}
+	err = tmr.snAccessor.SetLatestSequenceNumber(message)
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest sequence number")
+	}
+
+	// Get list of tasks on instance.
+	tasksToMessage, err := tmr.taskComparer.CompareRunningTasksOnInstanceWithManifest(message)
+	if err != nil {
+		return err
+	}
+
+	// Acknowledge the task manifest message asynchronously.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		tmr.ackTaskManifestMessage(message)
+	}()
+
+	// Send the task stop verification message to ACS asynchronously.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		tmr.sendTaskStopVerification(message, tasksToMessage)
+	}()
+	wg.Wait()
+	return nil
+}
+
+func (tmr *taskManifestResponder) ackTaskManifestMessage(message *ecsacs.TaskManifestMessage) {
+	messageID := aws.StringValue(message.MessageId)
+	logger.Info(fmt.Sprintf("Acking %s id: %s", TaskManifestMessageName, messageID))
+	err := tmr.respond(&ecsacs.AckRequest{
+		Cluster:           aws.String(tmr.cluster),
+		ContainerInstance: message.ContainerInstanceArn,
+		MessageId:         aws.String(messageID),
+	})
+	if err != nil {
+		logger.Error(fmt.Sprintf("Error Acking %s", TaskManifestMessageName), logger.Fields{
+			field.MessageID: messageID,
+			field.Error:     err,
+		})
+	}
+}
+
+func (tmr *taskManifestResponder) sendTaskStopVerification(message *ecsacs.TaskManifestMessage, tasksToStop []*ecsacs.TaskIdentifier) {
+	messageID := aws.StringValue(message.MessageId)
+	if len(tasksToStop) == 0 {
+		return
+	}
+	for _, task := range tasksToStop {
+		logger.Info("Sending task stop verification message", logger.Fields{
+			field.MessageID: messageID,
+			field.TaskARN:   task.TaskArn,
+		})
+	}
+	err := tmr.respond(&ecsacs.TaskStopVerificationMessage{
+		MessageId:      message.MessageId,
+		StopCandidates: tasksToStop,
+	})
+	if err != nil {
+		for _, task := range tasksToStop {
+			logger.Error("Error sending task stop verification message", logger.Fields{
```

Comment:
nit: Prefer to log with level WARN here instead of ERROR (for consistency with other responders)